### PR TITLE
WIP: Player repo

### PR DESCRIPTION
**Pull request by danielglasgow**

### CODE REVIEW

File: src/main/scala/server/CatanServer.scala [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        //Todo throw error if player already declared position
+        builders.put(req.gameId, gameContext.subscribePlayer(req.name, None, responseObserver))
+      }
+      case None => println("No game found with " + req.gameId + " cannot register listener: " + req.name)
+    }
+  }
+
+  override def move(req: MoveRequest) = {
+    val playerContext = games.get(req.gameId).flatMap(_.getPlayer(req.position)).getOrElse(throw new Exception(""))
+    playerContext.receiveMove(req.gameId, req.position, RollDiceMove)
```
Comment (danielglasgow) [inline src/main/scala/server/CatanServer.scala]: @grogdotcom This is the secrete sauce! Love it 😄

File: src/main/scala/soc/playerRepository/PlayerContext.scala [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        randomMoveSelector.moveRobberAndStealMove(state, inventory, position)
+      case MoveRequest(gameId, state: GameState[PLAYER], inventory: PerfectInfo, position, _) =>
+        randomMoveSelector.turnMove(state, inventory, position)
+
+      case _ =>
+        val observer: StreamObserver[GameUpdate] = gameObservers.getOrElse(playerKey, throw new Exception(""))
+        observer.onNext(new GameUpdate(Seq(name), request.toString, request.playerId))
+
+        val responsePromise = Promise.apply[CatanMove]()
+        expectedResponses.put(playerKey, responsePromise)
```
Comment (danielglasgow) [inline src/main/scala/soc/playerRepository/PlayerContext.scala]: So we create a promise but only resolve it when we get the response. Works for me!

File: src/main/scala/soc/playerRepository/PlayerContext.scala [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        responsePromise.future
+    }
+  }
+
+  def receiveMove(gameId: String, position: Int, move: CatanMove): Unit = {
+    expectedResponses.get((gameId, position)).map(_.success(move))
```
Comment (danielglasgow) [inline src/main/scala/soc/playerRepository/PlayerContext.scala]: ❤️

File: src/main/scala/soc/playerRepository/PlayerContext.scala [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+import soc.game.inventory.Inventory.PerfectInfo
+import soc.game.player.moveSelector.PossibleMoveSelector
+import soc.protos.game.GameUpdate
+import soc.storage.GameId
+
+import scala.collection.mutable.HashMap
+import scala.concurrent.{Future, Promise}
+import scala.util.Random
+
+class PlayerContext[GAME <: Inventory[GAME], PLAYER <: Inventory[PLAYER]](val name: String) {
```
Comment (danielglasgow) [inline src/main/scala/soc/playerRepository/PlayerContext.scala]: When do we create a new player context? 

It seems like right now there is a PlayerContext per "client strategy"?

I think we need a diagram to show the relationship between the GameBehavior, PlayerContext, and the clients.

File: src/main/scala/server/CatanServer.scala [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+      case None => println("No game found with " + req.gameId)
+    }
+    Future.successful(reply)
+  }
+
+  override def subscribe(req: SubscribeRequest, responseObserver: StreamObserver[GameUpdate]) = {
+    builders.get(req.gameId) match {
+      case Some(gameContext) => {
+        println("Registering listener " + req.name)
+        if (!playerRepo.contains(req.name)) playerRepo.addPlayer(new PlayerContext[PerfectInfo, NoInfo](req.name))
```
Comment (danielglasgow) [inline src/main/scala/server/CatanServer.scala]: Why not always create a new player context and pass in the observer--then below, attach the PlayerContext to a game context?

Whats the purpose of the player repo?

File: src/main/scala/soc/playerRepository/PlayerContext.scala (hunk 4 above)
Comment (grogdotcom) [reply src/main/scala/soc/playerRepository/PlayerContext.scala]: Yea I like the 1 to 1 nature of Player Context per "client strategy". It gives us an acces point to that particular client so if we ever do logins and accounts all that info has a place to go

Comment (danielglasgow) [reply src/main/scala/soc/playerRepository/PlayerContext.scala]: I think there should be a 1-1 relationship between PlayerContext and client. The server should have no knowledge of strategies versus clients. The idea that a single strategy can be playing multiple games is irrelevant to the server. It just knows each game has 4 players, each associated with a client stream to which it pushes updates. In other words, the server should have no conception of the player behind the client. 

IMO, the only responsibility of the GameServer is to play a game. The concept that certain "Players" are competing in multiple games or that a Player played yesterday, and is playing again today, is outside the scope of the GameServer. Baking those details in implicitly, by have clients registered as associated with players seems like client side implementation details bleeding across API boundaries.

**Pull request merged**